### DISCUSSION TITLE
console.log(watch()) breaks controlled form

### DISCUSSION THREAD

**Discussion by jvgeee** [react-hook-form/react-hook-form#3295] [Q&A] [answered]
**Describe the bug**
When trying to use a controlled form component with `useState`, if using `console.log(watch())` anywhere the form breaks. Omitting the console log makes things work as expected.

**To Reproduce**

See the sandbox, while console.log(watch()) is there you cannot type in the input, and if you try console logging within the `onChange` function it will never be called

**Codesandbox link (Required)**
https://codesandbox.io/s/modern-wood-kzi6c?file=/src/index.js

- [useForm](https://codesandbox.io/s/react-hook-form-useform-template-lsm61) Template
- [useFieldArray](https://codesandbox.io/s/react-hook-form-usefieldarray-template-q0jek) Template
- [Controller](https://codesandbox.io/s/react-hook-form-controller-template-yu0z5) Template

**Comment by bluebill1049** [comment]
you are mixing controlled input with uncontrolled. 

```tsx
import * as React from "react";
import * as ReactDOM from "react-dom";
import { useForm } from "react-hook-form";

import "./styles.css";

const App = () => {
  const { register, watch } = useForm();

  // Comment out this line to fix form
  console.log(watch());

  return (
    <div>
      <label>First Name</label>
      <input
        name="firstName"
        type="text"
        ref={register}
        pattern="[A-Za-z]+"
      />
    </div>
  );
};
```

**Comment by jvgeee** [comment]
Maybe I'm not understanding the concepts correctly. In actual use, it's part of a bigger form which has some elements controlled (I want to limit the input so that only letters can appear, afaik the only way to do this is using state + onChange) and others that are uncontrolled.

I reduced the form to only have one field so I could display the issue better, but maybe it's not useful.

See below; what am I missing here?


```
const App = () => {
  const { register, watch, handleSubmit, setError, errors } = useForm();
  const [fName, setFName] = React.useState("");

  // Comment out this line to fix form
  console.log(watch());

  const onChangeFName = (event) => {
    console.log("In onChange function");
    if (event.target.validity.valid) {
      setFName(event.target.value);
    }
  };

  return (
    <form>
    <div>
      <label>First Name</label>
      <input
        name="firstName"
        type="text"
        ref={register}
        onChange={onChangeFName}
        value={fName}
        pattern="[A-Za-z]+"
      />
    </div>
    <div>
      <label>Email</label>
      <input
        name="email"
        type="text"
        ref={register}
      />
    </div>
    <div>
      <label>password</label>
      <input
        name="password"
        type="password"
        ref={register}
      />
      </div>
    </form>
  );
};
```

**Comment by bluebill1049** [comment, marked as answer]
why do you need `setName`?

**Reply by bluebill1049** [reply]
if you need validation, then just register({ pattern: xxx })

**Reply by jvgeee** [reply]
Because my client wants the UI to only allow the user to type in valid inputs, so I'm using `onChange` to validate the content before it's allowed to update.

(the use-case might sound dumb, but it'll be used in many places, like a custom date of birth field that just accepts numbers)

**Reply by jvgeee** [reply]
I tried `register({pattern: xxx})`, but doesn't this validation only work once the form is submitted?

**Reply by bluebill1049** [reply]
if you want to prevent the user's input.

`<input onChnge={() => e.target.value = 'xxx'} ref={register} name="mask" />`

or event better considers using a mask input component.

**Reply by jvgeee** [reply]
Ah, gotcha. For anyone else who comes across this, here's a sandbox with `react-hook-from` working witih `react-input-mask`: https://codesandbox.io/s/react-hook-form-with-inputmask-d4fvd?file=/src/App.js:1009-1168

**Reply by bluebill1049** [reply]
thanks for sharing 👍
